feat: validate new issues GHA

iowillhoit opened the pull request:
Adds a new Github action that validates new issues. 
The Action will:
- Post a comment if `sfdx` or `sf` are older than latest
- Post a comment if the full `--verbose` version output was not included
- Add the `more information needed` label

Here is an example of all 3 cases: https://github.com/iowillhoit/gha-sandbox/issues/22

This PR also updates ISSUE_TEMPLATE

Testing Instructions:
- Create issues here: https://github.com/iowillhoit/gha-sandbox/issues
- Versions check looks for `@salesforce/cli@x.x.x` and `sfdx-cli/x.x.x`
- Verbose check looks for `pluginVersions` or `Plugin Version:` (to support json and stdout)

iowillhoit committed:
feat: validate new issues GHA

iowillhoit commented:
@jshackell-sfdc 3 markdown files for review here. The rendered templates can be viewed here: https://github.com/iowillhoit/gha-sandbox/issues/22

jshackell-sfdc committed:
edit bug_report.md

mshanemc commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -6,41 +6,39 @@ labels: 'investigating'
 
 ### System Information
+<!-- Which shell/terminal are you using? (bash, zsh, powershell 7, cmd.exe, etc) -->
+<!-- Paste the **full** output of the `version --verbose --json` command below -->
```
suggestion: does doctor provide anything better that the version --verbose doesn't that we might want instead or in addition?

iowillhoit replied:
Doctor does some version and plugin checks but it does not actually output any of it. I think we still want to suggest this for now so we get version, plugins, os, node version, etc

mshanemc commented on .github/actions/validate-issue/action.yml:
```diff
@@ -0,0 +1,17 @@
+    required: true
+    description: "Token taken from secrets env var"
+runs:
+  using: "node16"
```
I'd go higher since 16 is EOL in October.

I couldn't tell from the docs if there's a way to specify lts, but that would be best.

iowillhoit replied:
It's not supported yet 😑 https://github.com/actions/runner-images/issues/7002#issuecomment-1470047544
I tried a work around there but it did not seem to work. Going to wait it out and hopefully they will add support

jshackell-sfdc committed:
edit old-cli.md

jshackell-sfdc committed:
edit provide-version.md

jshackell-sfdc committed:
Update old-cli.md

jshackell-sfdc committed:
Update bug_report.md

mshanemc commented on .github/workflows/new-label.yml:
```diff
@@ -46,6 +46,21 @@ jobs:
           If you require immediate assistance, contact Salesforce Customer Support.
     - name: log action output
       run: echo ${{steps.run_action.outputs.message}}
+  validate-new-issue:
```
if I get your auto-response [say I forgot to put my version stuf], and then I modify the issue to include that, will this rerun?  I'm thinking **no** because it's triggered off of new labels.  

We'd probably want it to "maintain" the `more information required` label until the issue meets the criteria (it might not need to comment again if the previous comment still stands, but it could make sure the label stays on)?

Maybe it should be triggered by something else (create/edit of an issue ?)

mshanemc commented on .github/actions/validate-issue/index.js:
```diff
@@ -0,0 +1,121 @@
+        owner,
+        repo,
+        issue_number,
+        labels: ["more information needed"],
```
what removes the `more information needed` label?  Should it remove `investigating`

Would it be good if additional info from the user (provided it meets the criteria) would 
1. remove `more info`
2. add `investigating` back ?

I'm trying to think of the easiest way for the Trust project to be accurate more often

mshanemc commented on .github/actions/validate-issue/index.js:
```diff
@@ -0,0 +1,121 @@
+    function getFile(filename, replacements) {
+      const contents = fs.readFileSync(path.join(__dirname, filename), "utf8");
+      return Object.entries(replacements || {}).reduce((acc, [key, value]) => {
+        return acc.replace(new RegExp(key, "g"), value);
```
new js fun! 
```suggestion
        return acc.replaceAll(key, value);
```

FWIW, I wouldn't use reduce for this.  Just loop over the replacement key/value and call replaceAll, return the final,  since the `reduce` is mutating the contents anyway.

mshanemc commented on .github/actions/validate-issue/index.js:
```diff
@@ -0,0 +1,121 @@
+const execSync = require("child_process").execSync;
+const semver = require("semver");
+
+(async function() {
```
remember top-level await is allowed on node16+, so you could just define a function and then then call it.  😄

mshanemc reviewed (commented):
some code comments and process questions.

iowillhoit committed:
chore: add/remove labels, supports comments

iowillhoit commented:
Ok @mshanemc , updated this to add and remove labels. Also added another workflow that will watch for updates to the PR body and comments. I _think_ we have good coverage on scenarios that we will encounter. I will try to monitor this and make sure it is working as expected over the coming weeks. 

See the note about adding a `validated` label. 

Testing comments: https://github.com/iowillhoit/gha-sandbox/issues/32
Testing body edit: https://github.com/iowillhoit/gha-sandbox/issues/33
sf and sfdx: https://github.com/iowillhoit/gha-sandbox/issues/34
Testing comment edits: https://github.com/iowillhoit/gha-sandbox/issues/35

Edit:
It also will display multiple non-latest versions if you've shared them. Example: https://github.com/iowillhoit/gha-sandbox/issues/35#issuecomment-1585167460

mshanemc reviewed (approved):
seems good, thanks for thinking through the label states.

iowillhoit committed:
add date check

iowillhoit commented on .github/actions/validate-issue/index.js:
```diff
@@ -0,0 +1,168 @@
+      return;
+    }
+
+    // Temporary check to prevent this action from running on old issues
+    // This will prevent noise on tickets already being investigated
+    // This can be removed once the action has been running for a while
+    const creationDate = new Date(issue.created_at);
+    const cutoffDate = new Date("2023-06-11T00:00:00Z");
+    if (creationDate < cutoffDate) {
+      console.log("Issue was created before 6/11/2023, skipping");
+      return;
+    }
```
One more thing, I added a date check to prevent this from running on old issues. Here is an example of the check failing: https://github.com/iowillhoit/gha-sandbox/actions/runs/5245104855/jobs/9472071608#step:5:8
